src/gf.c

```diff
@@ -1796,125 +1719,127 @@ JL_DLLEXPORT jl_value_t *jl_matching_methods(jl_tupletype_t *types, int lim, int
+    }
 
-   if (jl_options.compile_enabled == JL_OPTIONS_COMPILE_OFF ||
-            jl_options.compile_enabled == JL_OPTIONS_COMPILE_MIN) {
+    if (jl_options.compile_enabled == JL_OPTIONS_COMPILE_OFF ||
+        jl_options.compile_enabled == JL_OPTIONS_COMPILE_MIN) {
         // copy fptr from the template method definition
-        jl_method_t *def = li->def.method;
+        jl_method_t *def = mi->def.method;
         if (jl_is_method(def) && def->unspecialized) {
-            jl_method_instance_t *unspec = def->unspecialized;
-            if (unspec->invoke != jl_fptr_trampoline) {
-                li->functionObjectsDecls.functionObject = NULL;
-                li->functionObjectsDecls.specFunctionObject = NULL;
-                li->specptr = unspec->specptr;
-                li->inferred = unspec->inferred;
-                if (li->inferred)
-                    jl_gc_wb(li, li->inferred);
-                li->inferred_const = unspec->inferred_const;
-                if (li->inferred_const)
-                    jl_gc_wb(li, li->inferred_const);
-                li->invoke = unspec->invoke;
-                return fptr;
+            jl_code_instance_t *unspec = def->unspecialized->cache;
+            if (unspec && unspec->invoke != NULL) {
+                jl_code_instance_t *codeinst = jl_set_method_inferred(mi, (jl_value_t*)jl_any_type, NULL, NULL,
+                    0, 1, ~(size_t)0);
+                codeinst->isspecsig = unspec->isspecsig;
+                codeinst->specptr = unspec->specptr;
+                codeinst->rettype_const = unspec->rettype_const;
+                codeinst->invoke = unspec->invoke;
+                return codeinst;
             }
         }
-        jl_code_info_t *src = jl_code_for_interpreter(li);
+        jl_code_info_t *src = jl_code_for_interpreter(mi);
         if (!jl_code_requires_compiler(src)) {
-            li->inferred = (jl_value_t*)src;
-            jl_gc_wb(li, src);
-            li->functionObjectsDecls.functionObject = NULL;
-            li->functionObjectsDecls.specFunctionObject = NULL;
-            li->invoke = jl_fptr_interpret_call;
-            return jl_fptr_interpret_call;
+            jl_code_instance_t *codeinst = jl_set_method_inferred(mi, (jl_value_t*)jl_any_type, NULL, NULL,
+                0, 1, ~(size_t)0);
+            codeinst->invoke = jl_fptr_interpret_call;
+            return codeinst;
         }
         if (jl_options.compile_enabled == JL_OPTIONS_COMPILE_OFF) {
             jl_printf(JL_STDERR, "code missing for ");
-            jl_static_show(JL_STDERR, (jl_value_t*)li);
+            jl_static_show(JL_STDERR, (jl_value_t*)mi);
             jl_printf(JL_STDERR, " : sysimg may not have been built with --compile=all\n");
         }
     }
 
-    jl_llvm_functions_t decls = li->functionObjectsDecls;
-    if (decls.functionObject == NULL) {
-        // if we don't have decls already, try to generate it now
+    codeinst = mi->cache;
+    while (codeinst) {
```

Comment:
Factor this lookup out into a separate function along with the others above that do a similar query?

Reply:
I'm not aware of any place that does the same lookup. C is a fairly terrible language, because the amount of work to combine this would seem likely to be more boilerplate work than just writing out the simple loop every time.